File: 05g-kafka-cloudevents-avro-with-springboot/src/spring-boot-cloudevents-kafka-producer-ce-native/src/main/java/com/trivadis/kafkaws/orders/OrderService.java
package com.trivadis.kafkaws.orders;

import com.example.orders.avro.OrderCreated;
import com.trivadis.kafkaws.events.CloudEventProducer;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class OrderService {

    private final CloudEventProducer cloudEventProducer;

    @Value("${app.kafka.topics.orders}")
    private String ordersTopic;

    public String placeOrder(String customerId, double amount, String currency) {
        String orderId = UUID.randomUUID().toString();

        // Build the generated Avro specific record
        OrderCreated event = OrderCreated.newBuilder()
            .setOrderId(orderId)
            .setCustomerId(customerId)
            .setAmount(amount)
            .setCurrency(currency)
            .setCreatedAt(Instant.now().toString())
            .build();

        cloudEventProducer.publish(
            ordersTopic,
            orderId,
            event,
            "com.example.orders.OrderCreated"
        );

        log.info("Order placed: {}", orderId);
        return orderId;
    }
}

File: 05g-kafka-cloudevents-avro-with-springboot/src/python-cloudevents-kafka-producer-ce-native/order_service.py
import uuid
from datetime import datetime, timezone
from pathlib import Path

from cloud_event_producer import CloudEventProducer

_SCHEMA_STR = (Path(__file__).parent / "avro" / "order-created.avsc").read_text()
_EVENT_TYPE = "com.example.orders.OrderCreated"
_SUBJECT    = "outbox.Order-value"


class OrderService:

    def __init__(self, producer: CloudEventProducer, topic: str):
        self._producer = producer
        self._topic = topic

    def place_order(self, customer_id: str, amount: float, currency: str) -> str:
        order_id = str(uuid.uuid4())

        payload = {
            "orderId":    order_id,
            "customerId": customer_id,
            "amount":     amount,
            "currency":   currency,
            "createdAt":  datetime.now(timezone.utc).isoformat(),
        }

        self._producer.publish(
            topic=self._topic,
            key=order_id,
            payload=payload,
            schema_str=_SCHEMA_STR,
            event_type=_EVENT_TYPE,
            subject_name=_SUBJECT,
        )

        print(f"Order placed: {order_id}")
        return order_id
